### ISSUE TITLE
produceSecureBindings() doesn't seem to work

### ISSUE THREAD

**Issue by asilism** [javadelight/delight-nashorn-sandbox#134]
Hi, recently I found that sandbox can load remote script by using load() function.
It can be secure problem in my project , so I checked option of sandbox.

I found allowLoadFunctions args is "false" by default.

I debugged NashornSandboxImpl, and there is ambiguous part of produceSecureBindings()

```java
private void produceSecureBindings() {
		try {
			final StringBuilder sb = new StringBuilder();
			cached = scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
			sanitizeBindings(cached);
			if (!allowExitFunctions) {
				sb.append("var quit=function(){};var exit=function(){};");
			}
			if (!allowPrintFunctions) {
				sb.append("var print=function(){};var echo = function(){};");
			}
			if (!allowReadFunctions) {
				sb.append("var readFully=function(){};").append("var readLine=function(){};");
			}
			if (!allowLoadFunctions) {
				sb.append("var load=function(){};var loadWithNewGlobal=function(){};");
			}
			if (!allowGlobalsObjects) {
				// Max 22nd of Feb 2018: I don't think these are strictly necessary since they
				// are only available in scripting mode
				sb.append("var $ARG=null;var $ENV=null;var $EXEC=null;");
				sb.append("var $OPTIONS=null;var $OUT=null;var $ERR=null;var $EXIT=null;");
			}
			scriptEngine.eval(sb.toString());

			resetEngineBindings();

			engineAsserted.set(true);

		} catch (final Exception e) {
			throw new RuntimeException(e);
		}
	}
```

It looks like doing method overide such as quit/load/print ... . but after that, call resetEngineBindings() again.
Doesn't it mean that all overrides are initialized?

I don't know if this is the real cause, but in my current sandbox it is still possible to execute load() function.
Is there something I'm missing?

**Comment by mxro** [comment]
Thank you for raising this issue!

I tried to reproduce it in a unit test:

```java
public class TestIssue134 {

	@Test
	public void test() throws ScriptCPUAbuseException, ScriptException {
		NashornSandbox sandbox = NashornSandboxes.create();

		sandbox.eval("load('classpath:TestIssue134.js')");
		sandbox.eval("load('somethingwrong')");
	}

}
```

https://github.com/javadelight/delight-nashorn-sandbox/commit/4f3c8faef7e40260dd4312cd12fccbb7ac3dd26d

It seems like the load function is just the overriden dummy.

When running the same test with the enabled load function as follows:

```java
	@Test
	public void test() throws ScriptCPUAbuseException, ScriptException {
		NashornSandbox sandbox = NashornSandboxes.create();
		sandbox.allowLoadFunctions(true);
		sandbox.eval("load('classpath:TestIssue134.js')");
		sandbox.eval("load('somethingwrong')");
	}
```

We will get the exception 'I should never be called' from running the script.

Could you help me refine the test so it captures what you have encountered? Thank you!

**Comment by asilism** [comment]
```java
@Test
	public void test() throws ScriptCPUAbuseException, ScriptException {
		NashornSandbox sandbox = NashornSandboxes.create();
		// create context for Engine scope binding
		ScriptContext context = new SimpleScriptContext();
		// actually, there is more source for binding
		// blah,blah, blah....
		sandbox.eval("load('classpath:TestIssue134.js')", context);
		sandbox.eval("load('somethingwrong')", context);
	}
```

I run eval with new sciprt context, but that is not secure context. 
Because the context isn't sanitize, and also allow/block function is based on context.
Is there any way to get sanitized context?

**Comment by mxro** [comment]
I think technically the Sandbox worked as expected here.

However, I think your example shows that it is very easy to accidentally expose scripts. Therefore, I released a new version 0.3.0 that attempts to make it safer working with script contexts.

It will now work as follows:

```java
        @Test
	public void test() throws ScriptCPUAbuseException, ScriptException {
		NashornSandbox sandbox = NashornSandboxes.create();
		// create context for Engine scope binding
		ScriptContext context = sandbox.createScriptContext();
		// actually, there is more source for binding
		// blah,blah, blah....
		sandbox.eval("load('classpath:TestIssue134.js')", context);
		sandbox.eval("load('somethingwrong')", context);
	}
```

Please let me know if this doesn't resolve your issue!

**Comment by mxro** [comment]
@dedaye Thank you, that's a good point on providing both bindings and script context.

What do you think of this solution?

```java
        if (bindings != null && scriptContext != null) {
            // Merge both: create a new context with scriptContext settings + merged engine bindings
            ScriptContext combinedContext = createMergedContext(scriptContext, bindings, scriptEngine);
            return scriptEngine.eval(js, combinedContext);
        } else if (bindings != null) {
```

https://github.com/javadelight/delight-nashorn-sandbox/pull/176/files#diff-2706d8a13b1a72512caa99c0d9fc5fd867aff96e751e0bc1a74ad899c3614a36R43

Feels a bit hacked together for me - but I cannot really think of something better. Maybe in addition we could mark the overloaded method with both bindings and script context as deprecated with a warning.

Relatedly, I have also been thinking on how to protect users from inadvertently providing an unsafe script context - something I think is easy to do.

I have been thinking of something along the lines of

```java
		if (scriptContext != null && !(scriptContext instanceof SandboxScriptContextImpl)) {
			LOG.warn("ScriptContext passed to eval is potentially dangerous. Use sandbox.createScriptContext() instead.");

		}
```
https://github.com/javadelight/delight-nashorn-sandbox/pull/176/files#diff-23b1d2e2dbf28976ed8d99118ac76094736e2bba6f6d4a3a2c4e10cdc8d610f8R533

but somehow didn't get it to work, and of course this is not really secure, but just more of a potential nudge.

**Comment by dedaye** [comment]
> [@dedaye](https://github.com/dedaye) Thank you, that's a good point on providing both bindings and script context.
> 
> What do you think of this solution?
> 
>         if (bindings != null && scriptContext != null) {
>             // Merge both: create a new context with scriptContext settings + merged engine bindings
>             ScriptContext combinedContext = createMergedContext(scriptContext, bindings, scriptEngine);
>             return scriptEngine.eval(js, combinedContext);
>         } else if (bindings != null) {
> https://github.com/javadelight/delight-nashorn-sandbox/pull/176/files#diff-2706d8a13b1a72512caa99c0d9fc5fd867aff96e751e0bc1a74ad899c3614a36R43
> 
> Feels a bit hacked together for me - but I cannot really think of something better. Maybe in addition we could mark the overloaded method with both bindings and script context as deprecated with a warning.
> 
> Relatedly, I have also been thinking on how to protect users from inadvertently providing an unsafe script context - something I think is easy to do.
> 
> I have been thinking of something along the lines of
> 
> 		if (scriptContext != null && !(scriptContext instanceof SandboxScriptContextImpl)) {
> 			LOG.warn("ScriptContext passed to eval is potentially dangerous. Use sandbox.createScriptContext() instead.");
> 
> 		}
> https://github.com/javadelight/delight-nashorn-sandbox/pull/176/files#diff-23b1d2e2dbf28976ed8d99118ac76094736e2bba6f6d4a3a2c4e10cdc8d610f8R533
> 
> but somehow didn't get it to work, and of course this is not really secure, but just more of a potential nudge.

I see that the javax.script.ScriptEngine class in JDK actually doesn't support a method with three parameters. Should we suggest adding a three-parameter method to the JDK? Before the JDK adds a three-parameter method, you can warn users who use the three-parameter eval method with appropriate warnings.

**Comment by mxro** [comment]
@dedaye  Thank you for having a look and for your thoughts.

I've added an additional warning message. I guess it could be quite disruptive if someone where to use it, but I think it is worth it, since before this release it would have been broken and after this release it would be bad for performance.

I've also added another warning message if someone provides an 'unsafe' script context, that would look like:

```
0    [main] WARN  delight.nashornsandbox.NashornSandbox  - ScriptContext passed to eval is potentially dangerous. Use sandbox.createScriptContext() instead to initialise the context.
```

Let me know if you think may be worthwhile to release.
